File: src/exoticatechnologies/modifications/exotics/impl/HangarForge.java
package exoticatechnologies.modifications.exotics.impl;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.combat.FighterLaunchBayAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.impl.campaign.ids.Stats;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.ui.UIComponentAPI;
import com.fs.starfarer.api.util.IntervalUtil;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;
import exoticatechnologies.modifications.exotics.Exotic;
import lombok.Getter;
import org.json.JSONObject;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;

public class HangarForge extends Exotic {
    private static final String ITEM = "et_hangarforge";
    private static final String REPLACEMENT_COUNT_ID = "et_fighterReplacements";
    private static final String REPLACEMENT_INTERVAL_ID = "et_fighterReplacementInterval";

    private static float RATE_DECREASE_MODIFIER = 25f;

    @Getter private final Color color = Color.GREEN;

    public HangarForge(String key, JSONObject settings) {
        super(key, settings);
    }

    @Override
    public boolean canAfford(CampaignFleetAPI fleet, MarketAPI market) {
        return Utilities.hasItem(fleet.getCargo(), ITEM);
    }

    @Override
    public boolean canApply(FleetMemberAPI member, ShipModifications mods) {
        if (member.getStats() != null) {
            if (member.getStats().getNumFighterBays().getModifiedInt() > 0) {
                return canApplyToVariant(member.getVariant());
            }
        }
        return false;
    }

    @Override
    public boolean removeItemsFromFleet(CampaignFleetAPI fleet, FleetMemberAPI fm, MarketAPI market) {
        Utilities.takeItemQuantity(fleet.getCargo(), ITEM, 1);
        return true;
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, MarketAPI market) {
        Map<String, Float> resourceCosts = new HashMap<>();
        resourceCosts.put(Utilities.formatSpecialItem(ITEM), 1f);
        return resourceCosts;
    }

    @Override
    public void modifyToolTip(TooltipMakerAPI tooltip, UIComponentAPI title, FleetMemberAPI fm, ShipModifications systems, boolean expand) {
        if (expand) {
            StringUtils.getTranslation(this.getKey(), "longDescription")
                    .format("rateDecreaseBuff", RATE_DECREASE_MODIFIER)
                    .addToTooltip(tooltip, title);
        }
    }

    @Override
    public void applyExoticToStats(FleetMemberAPI fm, MutableShipStatsAPI stats, float bandwidth, String id) {
        stats.getDynamic().getStat(Stats.REPLACEMENT_RATE_DECREASE_MULT).modifyMult(getBuffId(), 1f + RATE_DECREASE_MODIFIER / 100f);
    }

    @Override
    public void advanceInCombat(ShipAPI ship, float amount, float bandwidth) {
        int replacements = getFreeReplacements(ship);

        maintainStatus(ship,
                this.getBuffId(),
                StringUtils.getTranslation(this.getKey(), "statusText")
                        .format("freeReplacements", replacements)
                        .toStringNoFormats()
        );

        if (replacements < calculateMaxReplacements(ship)) {
            advanceReplacementInterval(ship, amount);
        }

        if (replacements > 0) {
            for (FighterLaunchBayAPI bay : ship.getLaunchBaysCopy()) {
                if (getDeadFighters(bay) > 0 && bay.getFastReplacements() == 0) {
                    bay.setFastReplacements(1);
                    bay.makeCurrentIntervalFast();
                    addFreeReplacements(ship, -1);
                    replacements--;

                    if (replacements == 0) break;
                }
            }
        }
    }

    public static int getDeadFighters(FighterLaunchBayAPI bay) {
        int dead = 0;

        if (bay.getWing() != null && bay.getWing().getWingMembers() != null) {
            dead += (getNumFighters(bay) - bay.getWing().getWingMembers().size());
            for (ShipAPI ship : bay.getWing().getWingMembers()) {
                if (!ship.isAlive()) {
                    dead++;
                }
            }
        }

        return dead;
    }

    public static int getNumFighters(FighterLaunchBayAPI bay) {
        return bay.getWing().getSpec().getNumFighters() + bay.getExtraDeployments();
    }

    public static void advanceReplacementInterval(ShipAPI ship, float amount) {
        IntervalUtil replacementInterval = getReplacementInterval(ship);
        replacementInterval.advance(amount);

        if (replacementInterval.intervalElapsed()) {
            addFreeReplacements(ship, calculateAddedReplacements(ship));
        }
    }

    private static IntervalUtil getReplacementInterval(ShipAPI ship) {
        if (!ship.getCustomData().containsKey(REPLACEMENT_INTERVAL_ID)) {
            ship.setCustomData(REPLACEMENT_INTERVAL_ID, new IntervalUtil(60f, 60f));
        }
        return (IntervalUtil) ship.getCustomData().get(REPLACEMENT_INTERVAL_ID);
    }

    private static int getFreeReplacements(ShipAPI ship) {
        if (!ship.getCustomData().containsKey(REPLACEMENT_COUNT_ID)) {
            int maxReplacements = calculateMaxReplacements(ship);
            setFreeReplacements(ship, maxReplacements);
        }
        return (int) ship.getCustomData().get(REPLACEMENT_COUNT_ID);
    }

    private static void setFreeReplacements(ShipAPI ship, int value) {
        ship.setCustomData(REPLACEMENT_COUNT_ID, value);
    }

    private static void addFreeReplacements(ShipAPI ship, int value) {
        setFreeReplacements(ship, getFreeReplacements(ship) + value);
    }

    private static int calculateAddedReplacements(ShipAPI ship) {
        return (int) Math.ceil(calculateMaxReplacements(ship) / 2f);
    }

    private static int calculateMaxReplacements(ShipAPI ship) {
        int wingCount = 0;
        for (String wingId : ship.getVariant().getWings()) {
            wingCount += Global.getSettings().getFighterWingSpec(wingId).getNumFighters();
        }

        return wingCount;
    }

    public void maintainStatus(ShipAPI ship, String id, String translation) {
        if(Global.getCombatEngine().getPlayerShip() != null
                && Global.getCombatEngine().getPlayerShip().equals(ship)) {
            Global.getCombatEngine().maintainStatusForPlayerShip(
                    id,
                    "graphics/icons/hullsys/reserve_deployment.png",
                    this.getName(),
                    translation,
                    false);
        }
    }
}

File: src/exoticatechnologies/modifications/exotics/impl/HangarForge.kt
package exoticatechnologies.modifications.exotics.impl

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.FighterLaunchBayAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.impl.campaign.ids.Stats
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.ui.UIComponentAPI
import com.fs.starfarer.api.util.IntervalUtil
import data.scripts.util.MagicUI
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.exotics.Exotic
import exoticatechnologies.modifications.exotics.ExoticData
import exoticatechnologies.modifications.exotics.types.ExoticType
import exoticatechnologies.util.RenderUtils
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities
import org.json.JSONObject
import java.awt.Color

class HangarForge(key: String, settings: JSONObject) : Exotic(key, settings) {
    override var color = Color.GREEN
    override fun canAfford(fleet: CampaignFleetAPI, market: MarketAPI): Boolean {
        return Utilities.hasItem(fleet.cargo, ITEM)
    }

    override fun canApply(member: FleetMemberAPI, mods: ShipModifications?): Boolean {
        if (member.stats != null) {
            if (member.stats.numFighterBays.modifiedInt > 0) {
                return canApplyToVariant(member.variant)
            }
        }
        return false
    }

    override fun removeItemsFromFleet(fleet: CampaignFleetAPI, member: FleetMemberAPI, market: MarketAPI): Boolean {
        Utilities.takeItemQuantity(fleet.cargo, ITEM, 1f)
        return true
    }

    override fun modifyToolTip(
        tooltip: TooltipMakerAPI,
        title: UIComponentAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData,
        expand: Boolean
    ) {
        if (expand) {
            StringUtils.getTranslation(key, "longDescription")
                .format("rateDecreaseBuff", RATE_DECREASE_MODIFIER)
                .addToTooltip(tooltip, title)
        }
    }

    override fun applyExoticToStats(
        id: String,
        stats: MutableShipStatsAPI,
        fm: FleetMemberAPI,
        mods: ShipModifications,
        data: ExoticData
    ) {
        stats.dynamic.getStat(Stats.REPLACEMENT_RATE_DECREASE_MULT)
            .modifyMult(buffId, 1f + RATE_DECREASE_MODIFIER / 100f)
    }

    override fun advanceInCombatAlways(ship: ShipAPI, member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData) {
        val replacements = getFreeReplacements(ship)
        val replacementInterval = getReplacementInterval(ship)
        if (replacementInterval != null) {
            MagicUI.drawInterfaceStatusBar(
                ship,
                replacementInterval.elapsed / replacementInterval.intervalDuration,
                RenderUtils.getAliveUIColor(),
                RenderUtils.getAliveUIColor(),
                0f,
                StringUtils.getString(key, "statusBarText"),
                replacements
            )
        }
    }

    override fun advanceInCombatUnpaused(
        ship: ShipAPI,
        amount: Float,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
        var replacements = getFreeReplacements(ship)
        if (Global.getCombatEngine().isPaused) {
            return
        }
        if (replacements < calculateMaxReplacements(ship)) {
            advanceReplacementInterval(ship, amount)
        }
        if (replacements > 0) {
            for (bay in ship.launchBaysCopy) {
                if (getDeadFighters(bay) > 0 && bay.fastReplacements == 0) {
                    bay.fastReplacements = 1
                    bay.makeCurrentIntervalFast()
                    addFreeReplacements(ship, -1)
                    replacements--
                    if (replacements == 0) break
                }
            }
        }
    }

    fun maintainStatus(ship: ShipAPI, id: String?, translation: String?) {
        if (Global.getCombatEngine().playerShip != null && Global.getCombatEngine().playerShip == ship) {
            Global.getCombatEngine().maintainStatusForPlayerShip(
                id,
                "graphics/icons/hullsys/reserve_deployment.png",
                name,
                translation,
                false
            )
        }
    }

    override fun canUseExoticType(type: ExoticType): Boolean {
        return false
    }

    companion object {
        private const val ITEM = "et_hangarforge"
        private const val REPLACEMENT_COUNT_ID = "et_fighterReplacements"
        private const val REPLACEMENT_INTERVAL_ID = "et_fighterReplacementInterval"
        private const val RATE_DECREASE_MODIFIER = 25f
        fun getDeadFighters(bay: FighterLaunchBayAPI): Int {
            var dead = 0
            if (bay.wing != null && bay.wing.wingMembers != null) {
                for (ship in bay.wing.wingMembers) {
                    if (!ship.isAlive) {
                        dead++
                    }
                }
            }
            return dead
        }

        fun advanceReplacementInterval(ship: ShipAPI, amount: Float) {
            val replacementInterval = getReplacementInterval(ship)
            replacementInterval!!.advance(amount)
            if (replacementInterval.intervalElapsed()) {
                addFreeReplacements(ship, calculateAddedReplacements(ship))
            }
        }

        private fun getReplacementInterval(ship: ShipAPI): IntervalUtil? {
            if (!ship.customData.containsKey(REPLACEMENT_INTERVAL_ID)) {
                ship.setCustomData(REPLACEMENT_INTERVAL_ID, IntervalUtil(60f, 60f))
            }
            return ship.customData[REPLACEMENT_INTERVAL_ID] as IntervalUtil?
        }

        private fun getFreeReplacements(ship: ShipAPI): Int {
            if (!ship.customData.containsKey(REPLACEMENT_COUNT_ID)) {
                val maxReplacements = calculateMaxReplacements(ship)
                setFreeReplacements(ship, maxReplacements)
            }
            return ship.customData[REPLACEMENT_COUNT_ID] as Int
        }

        private fun setFreeReplacements(ship: ShipAPI, value: Int) {
            ship.setCustomData(REPLACEMENT_COUNT_ID, value)
        }

        private fun addFreeReplacements(ship: ShipAPI, value: Int) {
            setFreeReplacements(ship, getFreeReplacements(ship) + value)
        }

        private fun calculateAddedReplacements(ship: ShipAPI): Int {
            return Math.ceil((calculateMaxReplacements(ship) / 2f).toDouble()).toInt()
        }

        private fun calculateMaxReplacements(ship: ShipAPI): Int {
            if (ship.variant == null || ship.variant.wings == null) {
                return 1
            }
            var wingCount = 0
            for (wingId in ship.variant.wings) {
                val spec = Global.getSettings().getFighterWingSpec(wingId)
                wingCount += spec?.numFighters ?: 1
            }
            return wingCount
        }
    }
}